src/modules/Workspaces/WorkspacesSnapshotTool/PwaHelper.cpp

```diff
@@ -0,0 +1,425 @@
+    };
+
+    struct IAppResolver_8 : public IUnknown
+    {
+    public:
+        virtual HRESULT STDMETHODCALLTYPE GetAppIDForShortcut() = 0;
+        virtual HRESULT STDMETHODCALLTYPE GetAppIDForShortcutObject() = 0;
+        virtual HRESULT STDMETHODCALLTYPE GetAppIDForWindow(HWND hWnd, WCHAR** pszAppId, void* pUnknown1, void* pUnknown2, void* pUnknown3) = 0;
+        virtual HRESULT STDMETHODCALLTYPE GetAppIDForProcess(DWORD dwProcessId, WCHAR** pszAppId, void* pUnknown1, void* pUnknown2, void* pUnknown3) = 0;
+    };
+
+    BOOL GetAppId_7(HWND hWnd, std::wstring* result)
+    {
+        HRESULT hr;
+
+        IAppResolver_7* AppResolver;
+        hr = CoCreateInstance(CLSID_StartMenuCacheAndAppResolver, NULL, CLSCTX_INPROC_SERVER | CLSCTX_INPROC_HANDLER, IID_IAppResolver_7, reinterpret_cast<void**>(&AppResolver));
+        if (SUCCEEDED(hr))
+        {
+            WCHAR* pszAppId;
+            hr = AppResolver->GetAppIDForWindow(hWnd, &pszAppId, NULL, NULL, NULL);
+            if (SUCCEEDED(hr))
+            {
+                *result = std::wstring(pszAppId);
+                CoTaskMemFree(pszAppId);
+            }
+
+            AppResolver->Release();
+        }
+
+        return SUCCEEDED(hr);
+    }
+
+    BOOL GetAppId_8(HWND hWnd, std::wstring* result)
+    {
+        HRESULT hr;
+        *result = L"";
+
+        IAppResolver_8* AppResolver;
+        hr = CoCreateInstance(CLSID_StartMenuCacheAndAppResolver, NULL, CLSCTX_INPROC_SERVER | CLSCTX_INPROC_HANDLER, IID_IAppResolver_8, reinterpret_cast<void**>(&AppResolver));
+        if (SUCCEEDED(hr))
+        {
+            WCHAR* pszAppId;
+            hr = AppResolver->GetAppIDForWindow(hWnd, &pszAppId, NULL, NULL, NULL);
+            if (SUCCEEDED(hr))
+            {
+                *result = std::wstring(pszAppId);
+                CoTaskMemFree(pszAppId);
+            }
+
+            AppResolver->Release();
+        }
+
+        return SUCCEEDED(hr);
+    }
+
+    BOOL PwaHelper::GetAppId(HWND hWnd, std::wstring* result)
+    {
+        HRESULT hr = GetAppId_8(hWnd, result);
+        if (!SUCCEEDED(hr))
+        {
+            hr = GetAppId_7(hWnd, result);
+        }
+        return SUCCEEDED(hr);
+    }
+
+    BOOL GetProcessId_7(DWORD dwProcessId, std::wstring* result)
+    {
+        HRESULT hr;
+        *result = L"";
+
+        IAppResolver_7* AppResolver;
+        hr = CoCreateInstance(CLSID_StartMenuCacheAndAppResolver, NULL, CLSCTX_INPROC_SERVER | CLSCTX_INPROC_HANDLER, IID_IAppResolver_7, reinterpret_cast<void**>(&AppResolver));
+        if (SUCCEEDED(hr))
+        {
+            WCHAR* pszAppId;
+            hr = AppResolver->GetAppIDForProcess(dwProcessId, &pszAppId, NULL, NULL, NULL);
+            if (SUCCEEDED(hr))
+            {
+                *result = std::wstring(pszAppId);
+                CoTaskMemFree(pszAppId);
+            }
```

Comment:
Please consider using smart pointers for strings, too.

```suggestion
            wil::unique_cotaskmem_string pszAppId;
hr = appResolver->GetAppIDForWindow(hWnd, &pszAppId, NULL, NULL, NULL);
if (SUCCEEDED(hr))
{
    *result = std::wstring(pszAppId.get());
}
```